### PR TITLE
Apply exception type and message

### PR DESCRIPTION
**Pull request by heyams**
testing is done [here](https://github.com/Azure-Samples/ApplicationInsights-Java-Samples/tree/main/opentelemetry-api/java-agent/TrackCustomException)

### CODE REVIEW

**Review by jeanbisutti — COMMENTED** [review]
Could you add automatic assertions on the new behavior?

**Comment by heyams** [comment]
smoke test didn't work. i'll wait for trask's feedback then.

**Comment by heyams** [comment]
note: customer is happy with this fix.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/AgentLogExporter.java [hunk 1]
```diff
@@ -111,7 +111,17 @@ public CompletableResultCode export(Collection<LogRecordData> logs) {
...
+            log.getAttributes()
+                .get(
+                    com.azure.monitor.opentelemetry.exporter.implementation.SemanticAttributes
+                        .EXCEPTION_TYPE);
+        String message =
+            log.getAttributes()
+                .get(
+                    com.azure.monitor.opentelemetry.exporter.implementation.SemanticAttributes
+                        .EXCEPTION_MESSAGE);
+        TelemetryItem telemetryItem = mapper.map(log, stack, type, message, itemCount);
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/AgentLogExporter.java:124]: (and move the code inside the method)
```suggestion
        TelemetryItem telemetryItem = mapper.map(log, stack, itemCount);
```

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/LogDataMapper.java [hunk 2]
```diff
@@ -151,7 +158,14 @@ private TelemetryItem createExceptionTelemetryItem(
     Attributes attributes = log.getAttributes();
     MAPPINGS.map(attributes, telemetryBuilder);
 
-    telemetryBuilder.setExceptions(Exceptions.minimalParse(stack));
+    ExceptionDetailBuilder exceptionDetailBuilder = Exceptions.minimalParse(stack).get(0);
```
Comment (trask) [inline agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/LogDataMapper.java:161]: how about change method to return a single:

```
public static ExceptionDetailBuilder minimalParse(String str)
```

Comment (trask) [reply agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/LogDataMapper.java:161]:
```suggestion
    List<ExceptionDetailBuilder> builders = Exceptions.minimalParse(stack);
    ExceptionDetailBuilder exceptionDetailBuilder = builders.get(0);
```

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/LogDataMapper.java [hunk 3]
```diff
@@ -151,7 +158,14 @@ private TelemetryItem createExceptionTelemetryItem(
...
+    if (type != null && !type.isEmpty()) {
+      exceptionDetailBuilder.setTypeName(type);
+    }
+    if (message != null && !message.isEmpty()) {
+      exceptionDetailBuilder.setMessage(message);
+    }
+    telemetryBuilder.setExceptions(Collections.singletonList(exceptionDetailBuilder));
```
Comment (trask) [inline agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/LogDataMapper.java:168]:
```suggestion
    telemetryBuilder.setExceptions(builders);
```

Comment (heyams) [reply agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/LogDataMapper.java:167]: exceptionDetailBuilder has the update.

**Comment by heyams** [comment]
@trask added smoke test, please check it out

File: smoke-tests/apps/OpenTelemetryApiLogBridge/src/main/java/com/microsoft/applicationinsights/smoketestapp/TestController.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  @GetMapping("/")
+  public String root() {
+    return "OK";
+  }
+
+  @GetMapping("/test-custom-exception-type-and-message")
+  public String testCustomExceptionTypeAndMessage() throws InterruptedException {
+    Span.current().updateName("myspanname");
+    SpanContext spanContext = SpanContext.create("ff01020304050600ff0a0b0c0d0e0f00", "090a0b0c0d0e0f00", TraceFlags.getSampled(), TraceState.getDefault());
+    try (Scope ignored = Span.wrap(spanContext).makeCurrent()) {
```
Comment (trask) [inline smoke-tests/apps/OpenTelemetryApiLogBridge/src/main/java/com/microsoft/applicationinsights/smoketestapp/TestController.java:32]: can be removed

File: smoke-tests/apps/OpenTelemetryApiLogBridge/src/main/java/com/microsoft/applicationinsights/smoketestapp/TestController.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      StringWriter sw = new StringWriter();
+      new Exception().printStackTrace(new PrintWriter(sw, true));
+
+      GlobalOpenTelemetry.get().getLogsBridge().get("my logger").logRecordBuilder()
+          .setSeverity(Severity.INFO)
+          .setAttribute(SemanticAttributes.EXCEPTION_TYPE, "my exception type")
+          .setAttribute(SemanticAttributes.EXCEPTION_MESSAGE, "This is an custom exception with custom exception type")
+          .setAttribute(SemanticAttributes.EXCEPTION_STACKTRACE, sw.toString())
+          .emit();
+      Thread.sleep(100000);
```
Comment (trask) [inline smoke-tests/apps/OpenTelemetryApiLogBridge/src/main/java/com/microsoft/applicationinsights/smoketestapp/TestController.java:42]:
```suggestion
```

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/LogDataMapper.java [hunk 6]
```diff
@@ -100,11 +102,12 @@ public LogDataMapper(
...
-  public TelemetryItem map(LogRecordData log, @Nullable String stack, @Nullable Long itemCount) {
+  public TelemetryItem map(LogRecordData log, @Nullable Long itemCount) {
+    String stack = log.getAttributes().get(SemanticAttributes.EXCEPTION_STACKTRACE);
     if (stack == null) {
       return createMessageTelemetryItem(log, itemCount);
     } else {
-      return createExceptionTelemetryItem(log, itemCount, stack);
+      return createExceptionTelemetryItem(log, itemCount);
     }
   }
```
Comment (trask) [inline agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/LogDataMapper.java:112]: `stack` was intentionally passed in as arg so we don't need to look it up multiple times

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/LogDataMapper.java (hunk 3 above)
Comment (trask) [reply agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/LogDataMapper.java:167]: can you ping me to discuss this? thx

File: smoke-tests/apps/OpenTelemetryApiLogBridge/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/OpenTelemetryApiLogBridgeTest.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    assertThat(rd.getProperties()).containsEntry("_MS.ProcessedByMetricExtractors", "True");
+
+    assertThat(rdEnvelope.getIKey()).isEqualTo("00000000-0000-0000-0000-0FEEDDADBEEF");
+    assertThat(rdEnvelope.getTags())
+        .hasEntrySatisfying("ai.internal.sdkVersion", v -> assertThat(v).startsWith("java:3."));
+
+    String operationId = rdEnvelope.getTags().get("ai.operation.id");
+    List<Envelope> edList =
+        testing.mockedIngestion.waitForItemsInOperation("ExceptionData", 1, operationId);
+    assertThat(edList.size()).isNotZero();
```
Comment (trask) [inline smoke-tests/apps/OpenTelemetryApiLogBridge/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/OpenTelemetryApiLogBridgeTest.java:53]: can we assert on more properties of the captured exception? (e.g. type name and message)

**Comment by heyams** [comment]
have you checked out the smoke test? what is the risk of not having smoke test for this?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/AgentLogExporter.java [hunk 8]
```diff
@@ -111,7 +111,7 @@ public CompletableResultCode export(Collection<LogRecordData> logs) {
           itemCount = Math.round(100.0 / samplingPercentage);
         }
 
-        TelemetryItem telemetryItem = mapper.map(log, stack, itemCount);
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/AgentLogExporter.java:114]: reorder to match prior (only to minimize changes)

**Review by trask — APPROVED** [review]

**Pull request merged**